docs/csharp/proto2.md

```diff
@@ -0,0 +1,184 @@
+    // Sets the value of the extension
+    void SetExtension<TValue>(Extension<T, TValue> extension, TValue value);
+    // Returns whether the extension is present in the message
+    bool HasExtension<TValue>(Extension<T, TValue> extension);
+    // Clears the value of the extension, removing it from the message
+    void ClearExtension<TValue>(Extension<T, TValue> extension);
+    // Clears the value of the repeated extension, removing it from the message. Calling GetExtension after this will always return null.
+    void ClearExtension<TValue>(RepeatedExtension<T, TValue> extension);
+}
+```
+
+### Extensions
+
+Extensions are generated in static containers like reflection classes and type classes. 
+For example for a file called `foo.proto` containing extensions in the file scope, a 
+`FooExtensions` class is created containing the extensions defined in the file scope.
+For easy access, this class can be used with `using static` to bring all extensions into scope.
+
+```proto
+option csharp_namespace = "FooBar";
+extend Foo {
+    optional Baz foo_ext = 124;
+}
+message Baz {
+    extend Foo {
+        repeated Baz repeated_foo_ext = 125;
+    }
+}
+```
+```cs
+public static partial class FooExtensions {
+    public static readonly Extension<Foo, Baz> FooExt = /* initialization */;
+}
+
+public partial class Baz {
+    public partial static class Extensions {
+        public static readonly RepeatedExtension<Foo, Baz> RepeatedFooExt = /* initialization */;
+    }
+}
+```
+```cs
+using static FooBar.FooExtensions;
+using static FooBar.Baz.Extensions;
+
+var foo = new Foo();
+foo.SetExtension(FooExt, new Baz());
+foo.GetOrInitializeExtension(RepeatedFooExt).Add(new Baz());
+```
+
+# APIs
+
+### Message initialization
+
+Checking message initialization is not handled automatically by the library, but can be done manually via the
+`IsInitialized` extension method in `MessageExtensions`. Please note, parsers and input streams don't check messages 
+for initialization on their own and throw errors. Instead it's up to you to handle messages with missing required fields
+in whatever way you see fit.
+
+### Extension registries
+
+Just like in Java, extension registries can be constructed to parse extensions when reading new messages
+from input streams. The API is fairly similar to the Java API with some added bonuses with C# syntax sugars.
+
+```proto
+message Baz {
+    extend Foo {
+        optional Baz foo_ext = 124;
+    }
+}
+```
+```cs
+var registry = new ExtensionRegistry() 
+{ 
+    Baz.Extensions.FooExt
+};
+var foo = Foo.Factory.WithExtensionRegistry(registry).ParseFrom(input);
+Assert.True(foo.HasExtension(Bas.Extensions.FooExt));
+var fooNoRegistry = Foo.Factory.ParseFrom(input);
+Assert.False(foo.HasExtension(Bas.Extensions.FooExt));
+```
+
+### Custom options
+
+The original `CustomOptions` APIs are now deprecated. Using the new generated extension identifiers, 
```

Comment:
optional nit: maybe add one more sentence of detail why deprecating CustomOptions is a good thing?